staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go

```diff
@@ -449,33 +458,57 @@ func (c *AvailableConditionController) deleteAPIService(obj interface{}) {
 		}
 	}
 	klog.V(4).Infof("Deleting %q", castObj.Name)
-	c.enqueue(castObj)
+	c.updateCacheForAPIService(castObj)
+	c.enqueueService(castObj)
 }
 
-// there aren't very many apiservices, just check them all.
-func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []*apiregistration.APIService {
+func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []string {
 	metadata, err := meta.Accessor(obj)
 	if err != nil {
 		utilruntime.HandleError(err)
 		return nil
 	}
+	c.RLock()
+	defer c.RUnlock()
+	if svcMap, ok := c.cache[metadata.GetNamespace()]; ok {
```

Comment:
`return c.cache[metadata.GetNamespace()][metadata.GetName()]` is simpler

Reply:
still outstanding

Reply:
If metadata.GetName() is not in cache, the shorter form would access nil pointer.
That's why I used the if check.

Reply:
map lookups on the zero value do not NPE. see https://play.golang.org/p/olUPHYDxYnP